File: go/wgs84.go
// SPDX-License-Identifier: BSD-3-Clause

// Package ndslivemath provides NDS.Live mathematical utilities for geographic
// tiling: WGS84 coordinate handling, Morton (Z-order) codes, packed tile IDs
// and NDS-coordinate bounding boxes.
//
// This is the Go port of the reference Python and C++ implementations. The
// public surface mirrors the Python package ndslive.math, adapted to idiomatic
// Go (errors instead of exceptions, multiple return values instead of tuples).
package ndslivemath

import (
	"fmt"
	"math"
)

// WGS84 / NDS conversion constants. These mirror the Python reference exactly.
const (
	// EarthRadiusInMeters is the approximate radius of Earth used for
	// great-circle (haversine) distance calculations.
	EarthRadiusInMeters = 6371000.8

	// MetersPerDegree is the approximate number of meters per degree of
	// latitude (and of longitude at the equator).
	MetersPerDegree = 111320.0
)

// LonNdsDelta and LatNdsDelta are the smallest representable angular steps in
// the NDS coordinate system. They are defined as package-level variables (not
// const) because they involve floating-point division.
var (
	// LonNdsDelta is 360 / (2^32 - 1).
	LonNdsDelta = 360.0 / float64((uint64(1)<<32)-1)
	// LatNdsDelta is 180 / (2^31 - 1).
	LatNdsDelta = 180.0 / float64((uint64(1)<<31)-1)
)

// Geometry-layer constants mirroring the C++ Wgs84<double> static members
// (cpp/include/ndsmath/wgs84.h). These intentionally use a power-of-two
// denominator (2^32 / 2^31) rather than LonNdsDelta / LatNdsDelta above (which
// use 2^32 - 1 / 2^31 - 1). The difference shows up in the 13th significant
// digit, but it is load-bearing for the geometry layer: Wgs84AABB antimeridian
// handling and tile-from-index construction must match the C++ reference
// bit-for-bit. Do NOT reuse LonNdsDelta / LatNdsDelta in the geometry layer.
var (
	// LonNdsDeltaPow2 is 360 / 2^32 (== 8.381903171539307e-08).
	LonNdsDeltaPow2 = 360.0 / math.Exp2(32)
	// LatNdsDeltaPow2 is 180 / 2^31 (== 8.381903171539307e-08, numerically
	// equal to LonNdsDeltaPow2).
	LatNdsDeltaPow2 = 180.0 / math.Exp2(31)
	// LonMin is the minimum longitude (-180).
	LonMin = -180.0
	// LonMax is the maximum longitude (180 - LonNdsDeltaPow2 == 179.99999991618097).
	LonMax = 180.0 - LonNdsDeltaPow2
	// LatMin is the minimum latitude (-90).
	LatMin = -90.0
	// LatMax is the maximum latitude (90 - LatNdsDeltaPow2).
	LatMax = 90.0 - LatNdsDeltaPow2
)

// Wgs84 represents a point on the Earth's surface using the WGS84 coordinate
// system. Lon (X) and Lat (Y) are in degrees; Alt (Z) is meters above the
// WGS84 ellipsoid.
type Wgs84 struct {
	Lon float64 // longitude in degrees, normalized into [-180, 180)
	Lat float64 // latitude in degrees, clamped into [-90, 90 - LatNdsDelta]
	Alt float64 // altitude in meters above the WGS84 ellipsoid
}

// NewWgs84 constructs a WGS84 point in degrees and normalizes it:
// longitude is wrapped into [-180, 180) and latitude is clamped into
// [-90, 90 - LatNdsDelta].
func NewWgs84(lon, lat, alt float64) Wgs84 {
	w := Wgs84{Lon: lon, Lat: lat, Alt: alt}
	w.Normalize()
	return w
}

// Normalize wraps longitude into [-180, 180) and clamps latitude into
// [-90, 90 - LatNdsDelta]. It is called automatically by NewWgs84; call it
// explicitly after mutating Lon or Lat directly.
//
// This mirrors the Python Wgs84.normalize() logic precisely, including the
// math.Mod (fmod) sign-preserving wrap and the boundary snapping for values
// very close to the positive limit.
func (w *Wgs84) Normalize() {
	// Use math.Mod (fmod) to preserve the sign, matching std::fmod / Python.
	w.Lon = math.Mod(w.Lon, 360.0)

	// Snap values very close to +180 down to 180 - LonNdsDelta.
	if math.Abs(w.Lon-(180.0-LonNdsDelta)) < LonNdsDelta {
		w.Lon = 180.0 - LonNdsDelta
	}

	// Wrap values outside [-180, 180).
	if w.Lon >= 180.0 {
		w.Lon -= 360.0
	} else if w.Lon < -180.0 {
		w.Lon += 360.0
	}

	// Latitude normalization.
	if math.Abs(w.Lat-(90.0-LatNdsDelta)) < LatNdsDelta {
		w.Lat = 90.0 - LatNdsDelta
	}
	if w.Lat > 90.0-LatNdsDelta {
		w.Lat = 90.0 - LatNdsDelta // clamp to max lat
	} else if w.Lat < -90.0 {
		w.Lat = -90.0 // clamp to min lat
	}
}

// ToNdsCoordinates converts this WGS84 point to NDS integer coordinates.
//
// NDS allows floor, truncate, or round for this conversion. Floor is used here
// (as recommended by NDS and matching the reference implementations). This is
// significant for negative coordinates, which floor toward negative infinity
// rather than truncating toward zero.
func (w Wgs84) ToNdsCoordinates() (int32, int32) {
	xNds := math.Floor((w.Lon / 360.0) * math.Exp2(32))
	yNds := math.Floor((w.Lat / 180.0) * math.Exp2(31))
	return int32(xNds), int32(yNds)
}

// Wgs84FromNdsCoordinates constructs a Wgs84 point (with Alt=0) from NDS
// integer coordinates. It is the inverse of ToNdsCoordinates.
func Wgs84FromNdsCoordinates(x, y int32) Wgs84 {
	lonMultiplier := 360.0 / math.Exp2(32)
	latMultiplier := 180.0 / math.Exp2(31)
	lon := float64(x) * lonMultiplier
	lat := float64(y) * latMultiplier
	return NewWgs84(lon, lat, 0.0)
}

// Wgs84FromMortonCode constructs a Wgs84 point (with Alt=0) from a Morton code.
//
// This mirrors the C++ Wgs84<double>::fromMortonCode
// (cpp/include/ndsmath/wgs84.h). Note that it scales BOTH the NDS x (longitude)
// and the NDS y (latitude) by the same factor 360 / 2^32 — unlike
// Wgs84FromNdsCoordinates, which scales latitude by 180 / 2^31. This difference
// is intentional in the reference and is relied upon by the geometry layer
// (e.g. the Wgs84AABB tile-from-index constructor); do not "fix" it.
func Wgs84FromMortonCode(mortonCode MortonCode) Wgs84 {
	bitScaling := 360.0 / math.Exp2(32)
	x, y := mortonCode.ToNdsCoordinates()
	return NewWgs84(float64(x)*bitScaling, float64(y)*bitScaling, 0.0)
}

// Longitude returns the longitude (X) in degrees. Mirrors the C++ accessor.
func (w Wgs84) Longitude() float64 { return w.Lon }

// Latitude returns the latitude (Y) in degrees. Mirrors the C++ accessor.
func (w Wgs84) Latitude() float64 { return w.Lat }

// Dx returns the X (longitude) component. Mirrors the C++ dx() accessor.
func (w Wgs84) Dx() float64 { return w.Lon }

// Dy returns the Y (latitude) component. Mirrors the C++ dy() accessor.
func (w Wgs84) Dy() float64 { return w.Lat }

// Sub returns the component-wise difference of this point and other, then
// re-normalizes the result (longitude wrap, latitude clamp). It mirrors the
// re-normalizing Wgs84 operator- of the C++/Python reference, which the
// ear-clipping triangulation relies on.
func (w Wgs84) Sub(other Wgs84) Wgs84 {
	return NewWgs84(w.Lon-other.Lon, w.Lat-other.Lat, 0.0)
}

// DegreesToMeters converts degree distances to meters at a given latitude.
//
// Longitude distance shrinks toward the poles (scaled by cos(latitude));
// latitude distance is constant. Returns (widthMeters, heightMeters).
func DegreesToMeters(lonDegrees, latDegrees, atLatitude float64) (float64, float64) {
	lonMeters := math.Abs(lonDegrees) * MetersPerDegree * math.Cos(degToRad(atLatitude))
	latMeters := math.Abs(latDegrees) * MetersPerDegree
	return lonMeters, latMeters
}

// NdsDistanceToMeters converts NDS coordinate distances to meters at a given
// latitude. Returns (widthMeters, heightMeters).
func NdsDistanceToMeters(ndsXDistance, ndsYDistance, atLatitude float64) (float64, float64) {
	lonDegrees := (ndsXDistance / math.Exp2(32)) * 360.0
	latDegrees := (ndsYDistance / math.Exp2(31)) * 180.0
	return DegreesToMeters(lonDegrees, latDegrees, atLatitude)
}

// DistanceTo returns the great-circle distance in meters to another point,
// computed via the haversine formula using EarthRadiusInMeters.
func (w Wgs84) DistanceTo(other Wgs84) float64 {
	dlat := degToRad(other.Lat - w.Lat)
	dlon := degToRad(other.Lon - w.Lon)
	a := math.Pow(math.Sin(dlat/2), 2) +
		math.Cos(degToRad(w.Lat))*math.Cos(degToRad(other.Lat))*math.Pow(math.Sin(dlon/2), 2)
	c := 2 * math.Atan2(math.Sqrt(a), math.Sqrt(1-a))
	return EarthRadiusInMeters * c
}

// BearingFrom returns the initial bearing (forward azimuth) in radians from
// other toward this point, measured clockwise from true north.
func (w Wgs84) BearingFrom(other Wgs84) float64 {
	lat1 := degToRad(w.Lat)
	lon1 := degToRad(w.Lon)
	lat2 := degToRad(other.Lat)
	lon2 := degToRad(other.Lon)
	y := math.Sin(lon2-lon1) * math.Cos(lat2)
	x := math.Cos(lat1)*math.Sin(lat2) - math.Sin(lat1)*math.Cos(lat2)*math.Cos(lon2-lon1)
	return math.Atan2(y, x)
}

// ToDegreeMinutesSeconds formats this point as degrees-minutes-seconds strings.
// It returns (latStr, lonStr) in the form "DD° MM' SS.ss\" N|S" and
// "DDD° MM' SS.ss\" E|W".
func (w Wgs84) ToDegreeMinutesSeconds() (string, string) {
	latStr := dms(math.Abs(w.Lat))
	if w.Lat < 0 {
		latStr += " S"
	} else {
		latStr += " N"
	}
	lonStr := dms(math.Abs(w.Lon))
	if w.Lon < 0 {
		lonStr += " W"
	} else {
		lonStr += " E"
	}
	return latStr, lonStr
}

// dms formats a non-negative degree value as "D° M' S.ss\"", matching the
// Python convert() helper which truncates degrees and minutes toward zero
// (int() of a non-negative value == floor).
func dms(value float64) string {
	degrees := int(value)
	minutes := int((value - float64(degrees)) * 60)
	seconds := (value - float64(degrees) - float64(minutes)/60.0) * 3600
	return fmt.Sprintf("%d° %d' %.2f\"", degrees, minutes, seconds)
}

// Equals reports whether two points have approximately equal longitude and
// latitude (relative tolerance 1e-12), mirroring Python's math.isclose.
func (w Wgs84) Equals(other Wgs84) bool {
	return isClose(w.Lon, other.Lon, 1e-12) && isClose(w.Lat, other.Lat, 1e-12)
}

func degToRad(deg float64) float64 {
	return deg * math.Pi / 180.0
}

// isClose mirrors Python's math.isclose with rel_tol and abs_tol=0.0.
func isClose(a, b, relTol float64) bool {
	return math.Abs(a-b) <= relTol*math.Max(math.Abs(a), math.Abs(b))
}

File: js/src/wgs84.ts
// SPDX-License-Identifier: BSD-3-Clause

import { MortonCode } from './morton.js';

/**
 * Represents a point on the Earth's surface using the WGS84 coordinate system.
 * Provides methods for coordinate conversion, normalization, and distance
 * calculations.
 *
 * This is a faithful port of the Python reference implementation
 * (`python/src/ndslive/math/wgs84.py`).
 */
export class Wgs84 {
  /** Approximate radius of Earth in meters. */
  static readonly EARTH_RADIUS_IN_METERS = 6371000.8;
  /** Longitude quantization step of the NDS coordinate grid (degrees). */
  static readonly LON_NDS_DELTA = 360 / (2 ** 32 - 1);
  /** Latitude quantization step of the NDS coordinate grid (degrees). */
  static readonly LAT_NDS_DELTA = 180 / (2 ** 31 - 1);

  // Geometry-layer constants mirroring the C++ `Wgs84<double>` static members
  // (`cpp/include/ndsmath/wgs84.h`). These intentionally use a power-of-two
  // denominator (`2^32` / `2^31`) rather than the `LON_NDS_DELTA` /
  // `LAT_NDS_DELTA` above (which use `2^32 - 1` / `2^31 - 1`). The difference
  // shows up in the 13th significant digit, but it is load-bearing for the
  // geometry layer: `Wgs84Aabb` antimeridian handling and tile-from-index
  // construction must match the C++ reference bit-for-bit. Do NOT reuse
  // `LON_NDS_DELTA` / `LAT_NDS_DELTA` in the geometry layer.
  /** Longitude quantization step using a power-of-two denominator. */
  static readonly LON_NDS_DELTA_POW2 = 360.0 / 2 ** 32; // == 8.381903171539307e-08
  /** Latitude quantization step using a power-of-two denominator. */
  static readonly LAT_NDS_DELTA_POW2 = 180.0 / 2 ** 31; // == 8.381903171539307e-08 (equal)
  /** Minimum representable longitude. */
  static readonly LON_MIN = -180.0;
  /** Maximum representable longitude (`180 - LON_NDS_DELTA_POW2`). */
  static readonly LON_MAX = 180.0 - Wgs84.LON_NDS_DELTA_POW2; // == 179.99999991618097
  /** Minimum representable latitude. */
  static readonly LAT_MIN = -90.0;
  /** Maximum representable latitude (`90 - LAT_NDS_DELTA_POW2`). */
  static readonly LAT_MAX = 90.0 - Wgs84.LAT_NDS_DELTA_POW2;

  /** Longitude in degrees, wrapped into `[-180, 180)`. */
  x: number;
  /** Latitude in degrees, clamped to `[-90, 90 - LAT_NDS_DELTA]`. */
  y: number;
  /** Altitude in meters above the WGS84 ellipsoid. */
  z: number;

  /**
   * Construct a WGS84 point in degrees.
   *
   * @param lon Longitude in degrees. Wrapped into `[-180, 180)`.
   * @param lat Latitude in degrees. Clamped to `[-90, 90)`.
   * @param alt Altitude in meters above the WGS84 ellipsoid.
   */
  constructor(lon = 0.0, lat = 0.0, alt = 0.0) {
    this.x = lon;
    this.y = lat;
    this.z = alt;
    this.normalize();
  }

  /** Longitude in degrees (the `x` component). Mirrors the C++ accessor. */
  longitude(): number {
    return this.x;
  }

  /** Latitude in degrees (the `y` component). Mirrors the C++ accessor. */
  latitude(): number {
    return this.y;
  }

  /** The `x` (longitude) component. Mirrors the C++ `dx()` accessor. */
  dx(): number {
    return this.x;
  }

  /** The `y` (latitude) component. Mirrors the C++ `dy()` accessor. */
  dy(): number {
    return this.y;
  }

  /**
   * Wrap longitude into `[-180, 180)` and clamp latitude into `[-90, 90)`.
   *
   * Called automatically by the constructor; call explicitly after mutating
   * `x` (longitude) or `y` (latitude) directly.
   */
  normalize(): void {
    const LON_NDS_DELTA = Wgs84.LON_NDS_DELTA;
    const LAT_NDS_DELTA = Wgs84.LAT_NDS_DELTA;

    // Use a sign-preserving modulo to match Python's math.fmod / C++ std::fmod.
    this.x = fmod(this.x, 360.0);

    // Snap values very close to the maximum representable longitude down.
    if (Math.abs(this.x - (180.0 - LON_NDS_DELTA)) < LON_NDS_DELTA) {
      this.x = 180.0 - LON_NDS_DELTA;
    }

    // Wrap into [-180, 180).
    if (this.x >= 180.0) {
      this.x -= 360.0;
    } else if (this.x < -180.0) {
      this.x += 360.0;
    }

    // Latitude normalization.
    if (Math.abs(this.y - (90.0 - LAT_NDS_DELTA)) < LAT_NDS_DELTA) {
      this.y = 90.0 - LAT_NDS_DELTA;
    }

    if (this.y > 90.0 - LAT_NDS_DELTA) {
      this.y = 90.0 - LAT_NDS_DELTA; // Clamp to max lat
    } else if (this.y < -90.0) {
      this.y = -90.0; // Clamp to min lat
    }
  }

  /**
   * Convert WGS84 coordinates to NDS integer coordinates.
   *
   * NDS spec allows floor, truncate, or round operations for this conversion.
   * Floor is used here as recommended by NDS for consistency with the tiling
   * scheme. Do NOT use truncation (`| 0`, `~~`, `Math.trunc`).
   *
   * @returns Tuple of `[xNds, yNds]` as integers.
   */
  toNdsCoordinates(): [number, number] {
    const xNds = Math.floor((this.x / 360.0) * 2 ** 32);
    const yNds = Math.floor((this.y / 180.0) * 2 ** 31);
    return [xNds, yNds];
  }

  /**
   * Construct a {@link Wgs84} point from NDS integer coordinates.
   * Inverse of {@link Wgs84.toNdsCoordinates}.
   *
   * @param x NDS longitude (signed 32-bit integer).
   * @param y NDS latitude (signed 31-bit integer).
   * @returns Wgs84 point with longitude / latitude in degrees and `alt = 0`.
   */
  static fromNdsCoordinates(x: number, y: number): Wgs84 {
    const lonMultiplier = 360.0 / 2 ** 32;
    const latMultiplier = 180.0 / 2 ** 31;
    const lon = x * lonMultiplier;
    const lat = y * latMultiplier;
    return new Wgs84(lon, lat);
  }

  /**
   * Construct a {@link Wgs84} point from a {@link MortonCode}.
   *
   * Mirrors the C++ `Wgs84<double>::fromMortonCode`
   * (`cpp/include/ndsmath/wgs84.h`). Note that this scales **both** the NDS x
   * (longitude) and the NDS y (latitude) by the same factor `360 / 2^32` —
   * unlike {@link Wgs84.fromNdsCoordinates}, which scales latitude by
   * `180 / 2^31`. This difference is intentional in the reference and is relied
   * upon by the geometry layer (e.g. the `Wgs84Aabb` tile-from-index
   * constructor); do not "fix" it.
   *
   * @param mortonCode A MortonCode encoding NDS integer coordinates.
   * @returns Wgs84 point with longitude / latitude in degrees and `alt = 0`.
   */
  static fromMortonCode(mortonCode: MortonCode): Wgs84 {
    const bitScaling = 360.0 / 2 ** 32;
    const [x, y] = mortonCode.toNdsCoordinates();
    return new Wgs84(x * bitScaling, y * bitScaling);
  }

  /**
   * Convert degree distances to meters at a given latitude.
   *
   * Longitude distance varies by latitude (shrinks toward poles); latitude
   * distance is constant.
   *
   * @param lonDegrees Longitude distance in degrees.
   * @param latDegrees Latitude distance in degrees.
   * @param atLatitude The latitude where measurement is taken (affects longitude distance).
   * @returns Tuple of `[widthMeters, heightMeters]`.
   */
  static degreesToMeters(
    lonDegrees: number,
    latDegrees: number,
    atLatitude: number,
  ): [number, number] {
    const METERS_PER_DEGREE = 111320.0;

    const lonMeters = Math.abs(lonDegrees) * METERS_PER_DEGREE * Math.cos(toRadians(atLatitude));
    const latMeters = Math.abs(latDegrees) * METERS_PER_DEGREE;

    return [lonMeters, latMeters];
  }

  /**
   * Convert NDS coordinate distances to meters at a given latitude.
   *
   * @param ndsXDistance X (longitude) distance in NDS units.
   * @param ndsYDistance Y (latitude) distance in NDS units.
   * @param atLatitude The latitude where measurement is taken.
   * @returns Tuple of `[widthMeters, heightMeters]`.
   */
  static ndsDistanceToMeters(
    ndsXDistance: number,
    ndsYDistance: number,
    atLatitude: number,
  ): [number, number] {
    const lonDegrees = (ndsXDistance / 2 ** 32) * 360.0;
    const latDegrees = (ndsYDistance / 2 ** 31) * 180.0;

    return Wgs84.degreesToMeters(lonDegrees, latDegrees, atLatitude);
  }

  /**
   * Format this point as degrees-minutes-seconds strings.
   *
   * @returns Tuple of `[latStr, lonStr]` in the form
   *   `"DD° MM' SS.ss\" N|S"` and `"DDD° MM' SS.ss\" E|W"`.
   */
  toDegreeMinutesSeconds(): [string, string] {
    const convert = (value: number): string => {
      const degrees = Math.trunc(value);
      const minutes = Math.trunc((value - degrees) * 60);
      const seconds = (value - degrees - minutes / 60) * 3600;
      return `${degrees}° ${minutes}' ${seconds.toFixed(2)}"`;
    };

    const lat = convert(Math.abs(this.y)) + (this.y < 0 ? ' S' : ' N');
    const lon = convert(Math.abs(this.x)) + (this.x < 0 ? ' W' : ' E');
    return [lat, lon];
  }

  /**
   * Great-circle distance to another point, computed via the haversine formula.
   *
   * @param other Another {@link Wgs84} point.
   * @returns Distance in meters along the Earth's surface
   *   (uses {@link Wgs84.EARTH_RADIUS_IN_METERS}).
   */
  distanceTo(other: Wgs84): number {
    const dlat = toRadians(other.y - this.y);
    const dlon = toRadians(other.x - this.x);
    const a =
      Math.sin(dlat / 2) ** 2 +
      Math.cos(toRadians(this.y)) * Math.cos(toRadians(other.y)) * Math.sin(dlon / 2) ** 2;
    const c = 2 * Math.atan2(Math.sqrt(a), Math.sqrt(1 - a));
    return Wgs84.EARTH_RADIUS_IN_METERS * c;
  }

  /**
   * Initial bearing (forward azimuth) from `other` toward this point.
   *
   * @param other Another {@link Wgs84} point — the starting point.
   * @returns Bearing in radians, measured clockwise from true north.
   */
  bearingFrom(other: Wgs84): number {
    const lat1 = toRadians(this.y);
    const lon1 = toRadians(this.x);
    const lat2 = toRadians(other.y);
    const lon2 = toRadians(other.x);
    const y = Math.sin(lon2 - lon1) * Math.cos(lat2);
    const x =
      Math.cos(lat1) * Math.sin(lat2) - Math.sin(lat1) * Math.cos(lat2) * Math.cos(lon2 - lon1);
    return Math.atan2(y, x);
  }

  /** Component-wise equality within a small relative tolerance (1e-12). */
  equals(other: Wgs84): boolean {
    return isClose(this.x, other.x) && isClose(this.y, other.y);
  }

  /** Component-wise addition of longitude/latitude (returns a new point). */
  add(other: Wgs84): Wgs84 {
    return new Wgs84(this.x + other.x, this.y + other.y);
  }

  /** Component-wise subtraction of longitude/latitude (returns a new point). */
  sub(other: Wgs84): Wgs84 {
    return new Wgs84(this.x - other.x, this.y - other.y);
  }

  /** Component-wise multiplication of longitude/latitude (returns a new point). */
  mul(other: Wgs84): Wgs84 {
    return new Wgs84(this.x * other.x, this.y * other.y);
  }

  /** Component-wise division of longitude/latitude (returns a new point). */
  div(other: Wgs84): Wgs84 {
    return new Wgs84(this.x / other.x, this.y / other.y);
  }

  toString(): string {
    return `Wgs84(lon=${this.x}, lat=${this.y})`;
  }
}

/** Sign-preserving floating-point modulo, equivalent to C/Python `fmod`. */
function fmod(a: number, b: number): number {
  return a - Math.trunc(a / b) * b;
}

function toRadians(deg: number): number {
  return (deg * Math.PI) / 180;
}

/** Mirror of Python's `math.isclose(a, b, rel_tol=1e-12)`. */
function isClose(a: number, b: number, relTol = 1e-12): boolean {
  return Math.abs(a - b) <= relTol * Math.max(Math.abs(a), Math.abs(b));
}
